src/shared/uart_framing/uart_framing.hpp

```diff
@@ -0,0 +1,217 @@
+     * @return a vector of bytes encoded with COBS
+     */
+    std::vector<uint8_t> cobsEncoding(const std::vector<uint8_t>& to_encode)
+    {
+        auto encoded = std::vector<uint8_t>();
+        encoded.emplace_back(START_END_FLAG_BYTE);
+
+        size_t overhead_location = 1;
+        uint8_t overhead         = 0x01;
+
+        for (auto byte : to_encode)
+        {
+            if (byte == START_END_FLAG_BYTE)
+            {
+                encoded.insert(encoded.begin() + overhead_location, overhead);
+                overhead_location = encoded.size();
+                overhead          = 0x01;
+            }
+            else
+            {
+                encoded.emplace_back(byte);
+                if (++overhead == 0xFF)
+                {
+                    encoded.insert(encoded.begin() + overhead_location, overhead);
+                    overhead_location = encoded.size();
+                    overhead          = 0x01;
+                }
+            }
+        }
+        encoded.insert(encoded.begin() + overhead_location, overhead);
+        encoded.emplace_back(START_END_FLAG_BYTE);
+
+        return encoded;
+    }
+
+    /**
+     * Implementation of modified Consistent Overhead Byte Stuffing(COBS) decoding
+     * https://en.wikipedia.org/wiki/Consistent_Overhead_Byte_Stuffing
+     *
+     * Uses a Delimiter byte at both the beginning(different) and end(normal)
+     *
+     * @param to_decode the vector of bytes to decode
+     * @return a vector of bytes decoded with COBS
+     */
+    std::optional<std::vector<uint8_t>> cobsDecoding(
+        const std::vector<uint8_t>& to_decode)
+    {
+        if (to_decode.front() != START_END_FLAG_BYTE ||
+            to_decode.back() != START_END_FLAG_BYTE)
+        {
+            return std::nullopt;
+        }
+        auto decoded = std::vector<uint8_t>();
+
+        uint16_t i = 1;
+        while (i < to_decode.size())
+        {
+            uint8_t overhead = to_decode[i++];
+            for (uint16_t j = 1; i < to_decode.size() && j < overhead; j++)
+            {
+                decoded.emplace_back(to_decode[i++]);
+            }
+            if (overhead < 0xFF && i != to_decode.size() - 1 &&
+                overhead != START_END_FLAG_BYTE)
+            {
+                decoded.emplace_back(START_END_FLAG_BYTE);
+            }
+        }
+        // Make sure decoded message is within the upper bound of bytes
+        if (decoded.size() + 3 + decoded.size() / 254 < to_decode.size())
+        {
+            return std::nullopt;
+        }
+        return decoded;
+    }
+}  // anonymous namespace
+
+/**
+ * Prepares a struct for being sent over uart.
+ * Performs both the framing and encoding.
+ *
+ * @tparam T type of struct beginning sent
+ * @param data struct being marshalled
+ * @return vector of bytes to be sent over uart
+ */
+template <typename T>
+std::vector<uint8_t> marshallUartPacket(const T& data)
+{
```

Comment:
thoughts on moving marshall and unmarshall to be functions of UartMessageFrame?

Reply:
Moving marshal to be a function of UartMessageFrame makes sense to me but unmarshalling doesn't because it returns an optional.